### DISCUSSION TITLE
CPT from causalImpact not adding up to 1

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#9] [Q&A] [answered]
How should I interpret this CPT when the entries do not add up to 1? I was running a randomBN with 32 nodes and `arc_ratio =3`. then I manually added 8 latent variables with at least 1 being connected to the variable X30. Is there a way to avoid this?

![Screenshot 2023-03-08 at 12 39 40 AM](https://user-images.githubusercontent.com/24616330/223628686-a7d16f6b-bc2a-422e-a870-b8dfa3ded8f5.png)

**Comment by phwuil** [comment]
Hi, Kenneth, I think we need a bit more information to answer that. Can you provide a minimal example ?

**Reply by kenneth-lee-ch** [reply]
@phwuil  It is pretty hard to generate the exact same bn by using `fastBN` since I generated the bn by using `randomBN()` with `n=32` and `ratio_arc = 3`. I saved it to a BIF file. I cannot upload it here due to BIF format, but I can attach a link and you can download it here:
https://drive.google.com/file/d/16IVhqGFEsjY4kyJZxYRjlXAXEQwJsGnZ/view?usp=sharing

```
bn_no_latent =gum.loadBN("bnfile.bif")
bn_latent_spec = [('U0', ('X3', 'X24')), ('U1', ('X29', 'X30')), ('U2', ('X1', 'X22')), ('U3', ('X18', 'X23')), ('U4', ('X22', 'X30')), ('U5', ('X5', 'X26')), ('U6', ('X0', 'X20')), ('U7', ('X13', 'X29'))]
cm2 = csl.CausalModel(bn_no_latent, bn_latent_spec,  keepArcs= True)
_, prob, _ = csl.causalImpact(cm2, on = set(['X30', 'X1', 'X22', 'X29']), doing= set(['X3', 'X20', 'X18', 'X4', 'X21', 'X23', 'X28', 'X24', 'X17', 'X11', 'X12']))
prob
```

This is what I see after running the above code:
![Screenshot 2023-03-15 at 2 58 32 PM](https://user-images.githubusercontent.com/24616330/225414839-652bc894-3bc9-4e8e-81a6-dba1ba90aae8.png)

**Comment by phwuil** [comment, marked as answer]
Hello @kenneth-lee-ch , the result is correct : your `on = set(['X30', 'X1', 'X22', 'X29'])`  expresses that you want to compute the joint distribution over X30, X1, X22, X29 , conditioned by interventions on 'X3', 'X20', 'X18', 'X4', 'X21', 'X23', 'X28', 'X24', 'X17', 'X11', 'X12'.

![image](https://user-images.githubusercontent.com/5980636/225652587-be7f9b99-60bb-4e31-8286-6788ca1ad7c3.png)


So it does sum to 1 but over all variables of `set(['X30', 'X1', 'X22', 'X29'])`.

**Reply by kenneth-lee-ch** [reply]
@phwuil  Then, I am a bit confused. Shouldn't the table of 
`P(X30| do(X3, X20, X18, X4, X21, X23, X28, X24, X17, X11, X12), X1, X22, X29)` be the same as that of
`P(X30, X1, X22, X29| do(X3, X20, X18, X4, X21, X23, X28, X24, X17, X11, X12))` according to Theorem 20 from llya's paper: (https://www.jmlr.org/papers/volume9/shpitser08a/shpitser08a.pdf)? 

There are backdoor paths from each of `X1, X22, X29` to `X30` according to the graph I exported from the causal model (attached here).

<img width="957" alt="Screenshot 2023-03-16 at 1 48 24 PM" src="https://user-images.githubusercontent.com/24616330/225708406-5d6378e6-7c5c-4a25-a25f-38ffbc82f672.png">

I also tried out the exact same bn with the query `P(X30| do(X3, X20, X18, X4, X21, X23, X28, X24, X17, X11, X12), X1, X22, X29)` 

```
_, prob, _ = csl.causalImpact(cm2, on = set(['X30']), doing= set(['X3', 'X20', 'X18', 'X4', 'X21', 'X23', 'X28', 'X24', 'X17', 'X11', 'X12']), knowing= set(['X1', 'X22', 'X29']))
prob
```
It produces the following table and clearly, it is different than `P(X30, X1, X22, X29| do(X3, X20, X18, X4, X21, X23, X28, X24, X17, X11, X12))`

<img width="1225" alt="Screenshot 2023-03-16 at 2 06 00 PM" src="https://user-images.githubusercontent.com/24616330/225712289-40439371-199c-4f23-95ca-e3058f710e69.png">

**Reply by phwuil** [reply]
Theorem 20 shows a way to swap between intervention do(z) and observation (given z) in certain cases. 

But note that it is always P...(y|....) : y does not change. In your formula, `P(X30|...)` becomes `P(X30, X1, X22, X29|...)`. So you change the Y of the theorem 20. It is not the theorem 20. 

If you sum out  X1, X22, X29 in the first code, I guess you will find the result of your second code.

**Reply by kenneth-lee-ch** [reply]
Oh, I see. I overlooked it. Thanks!
